pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java

```diff
@@ -152,44 +179,87 @@ private TransferableBlock buildJoinedDataBlock(TransferableBlock leftBlock)
     if (leftBlock.isErrorBlock()) {
       _upstreamErrorBlock = leftBlock;
       return _upstreamErrorBlock;
-    } else if (TransferableBlockUtils.isNoOpBlock(leftBlock) || TransferableBlockUtils.isEndOfStream(leftBlock)) {
+    }
+    if (leftBlock.isNoOpBlock() || (leftBlock.isSuccessfulEndOfStreamBlock() && !needUnmatchedRightRows())) {
       return leftBlock;
     }
+    // TODO: Moved to a different function.
+    if (leftBlock.isSuccessfulEndOfStreamBlock() && needUnmatchedRightRows()) {
+      // Return remaining non-matched rows for non-inner join.
+      List<Object[]> returnRows = new ArrayList<>();
+      for (Map.Entry<Key, List<Object[]>> entry : _broadcastRightTable.entrySet()) {
+        Set<Integer> matchedIdx = _matchedRightRows.getOrDefault(entry.getKey(), new HashSet<>());
+        List<Object[]> rightRows = entry.getValue();
+        if (rightRows.size() == matchedIdx.size()) {
+          continue;
+        }
+        for (int i = 0; i < rightRows.size(); i++) {
+          if (!matchedIdx.contains(i)) {
+            returnRows.add(joinRow(null, rightRows.get(i)));
+          }
+        }
+      }
+      _isTerminated = true;
+      return new TransferableBlock(returnRows, _resultSchema, DataBlock.Type.ROW);
+    }
     List<Object[]> rows = new ArrayList<>();
     List<Object[]> container = leftBlock.isEndOfStreamBlock() ? new ArrayList<>() : leftBlock.getContainer();
     for (Object[] leftRow : container) {
+      Key key = new Key(_leftKeySelector.getKey(leftRow));
       // NOTE: Empty key selector will always give same hash code.
-      List<Object[]> hashCollection =
-          _broadcastHashTable.getOrDefault(new Key(_leftKeySelector.getKey(leftRow)), Collections.emptyList());
-      // If it is a left join and right table is empty, we return left rows.
-      if (hashCollection.isEmpty() && _joinType == JoinRelType.LEFT) {
-        rows.add(joinRow(leftRow, null));
-      } else {
-        // If it is other type of join.
-        for (Object[] rightRow : hashCollection) {
-          // TODO: Optimize this to avoid unnecessary object copy.
-          Object[] resultRow = joinRow(leftRow, rightRow);
-          if (_joinClauseEvaluators.isEmpty() || _joinClauseEvaluators.stream().allMatch(evaluator ->
-              (Boolean) FunctionInvokeUtils.convert(evaluator.apply(resultRow), DataSchema.ColumnDataType.BOOLEAN))) {
-            rows.add(resultRow);
+      List<Object[]> matchedRightRows = _broadcastRightTable.getOrDefault(key, null);
+      if (matchedRightRows == null) {
+        if (needUnmatchedLeftRows()) {
+          rows.add(joinRow(leftRow, null));
+        }
+        continue;
+      }
```

Comment:
suggest clean up for the logic. no need to add join rows separately for equality or inequality in 2 places
```suggestion
      boolean hasMatchForLeftRow = false;
      if (matchedRightRows != null) {
        // ...
      } 
      if (!hasMatchForLeftRow && needUnmatchedLeftRows()) {
        rows.add(joinRow(leftRow, null));
      }
```